Step4

Mo-bile opened the pull request:
안녕하세요!
step 4 PR 요청드립니다 

이전에 피드백 주신 바 대로 Process에 책임을 덜어내기 위한 리팩토링에 집중을 하고나서, 우승자를 가리는 기능에 집중했습니다.

# 질문
## 1. 규칙6의 정량적인 느낌을 잡기가 어렵습니다
이전에 정량적 원칙으로 생각하면 좋은 것으로 알려주신 객체지향 생활체조 9원칙에 의거하여 개발하고 리팩토링 하려고했습니다.
그런데 규칙중 '규칙 6: 모든 엔티티를 작게 유지한다.' 기준을 명확히 잡기가 어렵네요 정성적인 느낌도 있어서요

## 2. 규칙3의 포장이 이중,삼중,사중이 되어도 괜찮은가요?
그리고 이차원리스트 이용을 지양하는 것을 피드백 주셨는데요 '규칙 3: 모든 원시값과 문자열을 포장한다.' 인데 이 규칙에 의거한것이라 생각했습니다. 이렇게 시도를 하였는데요
느낀점은 포장에 포장에 포장에 포장 즉 이중,삼중,사중으로 포장이 된 느낌이었습니다.
이런방식이 바람직한걸까요?

## 3. model의 필드값을 행동으로 리턴하게 하는 방식이 아직 어려운것 같습니다 좋은 방법이 있을까요?
record 자료형이 아닌 이상 직접 조회가 어렵네요. 
이 부분에 대해서 고민을 하면서 행동으로 데이터 조회하게하자! 생각하면할수록 클래스가 복잡해 지네요. 반대로 이부분을 포기하니 단순해지고요. 이 부분의 여부가 저에게 설계의 난이도 차이가 체감됩니다.

---

# 개인적으로 아쉬운 코드 (리팩토링이 필요하다고 생각하는 부분)
- `ProgressRecord.findWinners()` 메서드 분리 혹은 일부 기능 다른 클래스로 이관
- 'GameMove.moveCar()' 메서드 분리 혹은 일부 기능 다른 클래스로 이관
- 포장에 포장을 거듭한 방식에 대한 아쉬움
  - `Integer,String` -> `ProgressRecord` -> `GameResult` -> `GameFinalResult`

---
# 나가며
평일이 되니 주말만큼 시간을 할당할 수가 없어서 이번 PR에 대한 퀄리티가 내심 아쉽습니다.
일단 여기까지 PR드리고 해주신 피드백을 보고 반영하겠습니다.
질문에 대해서 꼼꼼하게 피드백 주셔서 감사드립니다.

Commit:
add dependency junit-jupiter-params

Mo-bile committed:
Merge branch 'next-step:mo-bile' into mo-bile

Mo-bile committed:
Merge branch 'next-step:mo-bile' into mo-bile

Mo-bile committed:
Merge branch 'next-step:mo-bile' into mo-bile

Mo-bile committed:
refactor : Car 클래스에 forward 와 isForwardByRandom 통합하여 Process 클래스가 이동에 필요한 조건을 모르게 하여 게임 진행 역할에만 집중하게됨

Mo-bile committed:
chore : bin/* 을 추가하여 빌드된 파일 git에 안나오게 반영

Mo-bile committed:
refactor
1. reules 의 car 필드를 제거하여 CarFactory에 이관하여 자동차 생성 한도 검증과 자동차 생성 관련 기능으로 변경

Mo-bile committed:
refactor
1. rules의 남은 moves 를 새로운 GameMove로 이관
2. GameMove의 역할은 게임 진행동안 move 입력값 검증, 이동진행, 각 자동차별 이동진행 역할을 한다.

Mo-bile committed:
refactror
1. showResult() 를 main으로 빼낸다. main 함수 역할은 결과 게임 결과에 대한 정리를 출력을 프리젠테이션 계층으로서 생각했기 때문
2. moveCar 에 대한 이동과 기록 정리까지 move에 이동시켜서 기록하게 함
3. Process에 단순하게 인스턴스 호출만 하게 함
4. 랜덤때문에 어려운 테스트를 단순 사이즈 정도로 넣어줌

Mo-bile committed:
fix : moveCar() 내 recordCarsMove() 위치를 수정하여 이동전에 먼저 자동차 기록하게 변경

Mo-bile committed:
doc : step4에 대한 설명 및 기능목록 추가

Mo-bile committed:
feat : 각 자동차에 이름 부여 완료

Mo-bile committed:
feat : 기록 필드에 name 추가

Mo-bile committed:
feat : 최종 출력 방식 반영

Mo-bile committed:
tmp : 임시커밋

Commit:
feat : 우승자 찾기 기능 개발 완료

Commit:
test : findWinnersTest 테스트 생성

javajigi commented on src/main/java/racingGame/business/CarFactory.java:
```diff
@@ -3,21 +3,23 @@
-        while(rules.isLimitCars(this.cars.size())) {
-            this.cars.add(new Car());
+
+    public List<Car> createCars(String[] names) {
```
```suggestion
    public static List<Car> createCars(String[] names) {
```
클래스 메서드로 구현하는 것은 어떨까?

javajigi commented on src/main/java/racingGame/business/GameMove.java:
```diff
@@ -0,0 +1,52 @@
+        this.moves = moves;
+    }
+
+    public GameFinalResult moveCar(List<Car> joinCars) {
```
전체 히스토리를 관리하게 될 경우 round 수가 많아질수록 레이싱 경주 동안 차지하는 메모리 크기가 커질 수 있지 않을까요?
각 round의 ProgressRecord이 단순히 로그를 위한 출력용이라면 히스토리 전체를 관리하기 보다 마지막 레이스 결과만 가지도록 구현해 보는 것은 어떨까요?
이렇게 구현할 경우 response에 있는 객체의 필요성도 사라질 수도 있을 것 같고, 테스트 측면에서도 좀 더 깔끔해지는 것도 있을 것 같아요

Mo-bile replied:
@javajigi 
제가 전체 히스토리 관리를 하게 한 이유는 UI 영역인 ResultView 로 전달은 한번만 가능해서라고 생각하기 때문입니다.

매 게임마다 전달이 가능한 경우는 `Process.run()` 에서 ResultView() 호출해야하는 경우로 보이기 때문입니다. 이 방식은 비즈니스 로직에 ui 출력관련 코드를 호출하는 방식같아서요

즉 Process가 UI를 모르고 데이터를 전달하게 하고싶은데, 그러련은 각 라운드별 게임의 히스토리를 모두 모아서 한번에 전달이 필요한게 아닐까요?

javajigi commented on src/main/java/racinggame/model/Car.java:
```diff
@@ -1,24 +1,37 @@
-    public int forward() {
-        return location++;
+
+    public int forward(int randomNum) {
```
👍

javajigi commented on src/main/java/racingGame/model/ProgressRecord.java:
```diff
@@ -5,15 +5,38 @@
 
 public class ProgressRecord {
     private final List<Integer> joinCarsRecord = new ArrayList<>();
+    private final List<String> joinCarsName = new ArrayList<>();
```
Car가 name과 location 데이터를 가지고 있는데 굳이 이 데이터를 밖으로 꺼내 위와 같이 구현할 필요가 있을까요?
name과 location 데이터 관련한 로직 구현이 필요한 Car에 접근해 구현하면 되지 않을까요?
위와 같이 구현하면 오히려 복잡도가 높아질 것 같은 느낌이 드네요.

javajigi commented on src/main/java/racingGame/model/ProgressRecord.java:
```diff
@@ -5,15 +5,38 @@
-        return this.joinCarsRecord;
+    }
+    
+    public List<String> findWinners() {
```
메서드가 너무 많은 책임을 맡고 있는 것은 아닐까요?
단위 책임 원칙을 지키도록 리팩터링해보면 좋겠음

javajigi commented on src/main/java/racingGame/Business/GameMove.java:
```diff
@@ -0,0 +1,52 @@
+package racingGame.Business;
```
모든 패키지 이름은 소문자로 구현하는 것이 컨벤션
여러 곳에서 대문자로 구현한 부분이 보이는데 모두 소문자로 리네임 리팩터링한다.

javajigi commented on src/test/java/racingGame/business/CarFactoryTest.java:
```diff
@@ -1,22 +1,35 @@
+    void carMakeTest() {
+        String[] strings = {"pobi", "crong","honux"};
+        carFactory = new CarFactory();
+        assertThat(carFactory.createCars(strings).size()).isEqualTo(3);
```
```suggestion
        assertThat(carFactory.createCars(strings)).hasSize(3);
```

javajigi commented on src/test/java/racingGame/model/ProgressRecordTest.java:
```diff
@@ -23,20 +24,53 @@ void recordIsEmptyTest() {
+        progressRecord.recordGame(joinCars);
         
-        assertThat(progressRecord.recordGame(joinCarsRecord).size()).isEqualTo(3);
+        assertThat(progressRecord.findWinners().size()).isGreaterThanOrEqualTo(1);
```
어떤 Car가 우승자인지를 테스트 데이터 초기화를 통해 구현하고 해당 우승자가 반환되는지 검증하도록 개선해 볼 수 있을까요?
지금은 단순히 크기만을 체크하는데 의미있는 테스트가 되려면 테스트에서 지정한 Car가 제대로 반환되는지 테스트할 수 있어야 더 의미있는 테스트가 될 것 같아요.

javajigi reviewed (changes requested):
4단계 구현하느라 수고했어요. 👍 
생각보다 쉽지 않죠?
피드백에도 남겼는데요.
자동차 경주 게임 전체 히스토리까지 관리하는 방식으로 구현하다보니 복잡도가 높아져 구현에 어려움이 있었을 것 같아요.
좀 더 단순한 구조로 접근해 봐도 좋지 않을까 생각합니다.
아직 객체 설계 역량이 부족한 상태에서 많은 객체를 만들고, 협력하다보니 오히려 구현에 어려움이 많이 느껴지네요.
불필요한 객체들 제거해봐도 좋지 않을까요?

Q1. 규칙6의 정량적인 느낌을 잡기가 어렵습니다
A1. 이 규칙은 정성적일 수 밖에 없고, 앞으로 의식적인 연습을 통해 감을 기를 수 밖에 없는 부분이라 생각해요.

Q2. 규칙3의 포장이 이중,삼중,사중이 되어도 괜찮은가요?
객체를 포장하다보면 2중, 3중 계속 이어질 겁니다.
제대로 OOP를 한다면 종종 발생하는 구조입니다.

Q3. model의 필드값을 행동으로 리턴하게 하는 방식이 아직 어려운것 같습니다 좋은 방법이 있을까요?
상태 값을 가지는 객체의 값에 직접 접근하거나 꺼내는(get) 방식으로 로직을 구현하지 말고 객체에 메시지를 보내 로직을 구현하는 방식으로 접근해 보세요.
OOP의 객체는 단순히 데이터 값을 가지는 수동적인 존재가 아니라 행위(로직)를 담당하는 능동적인 객체입니다. 이 부분이 다소 추상적일텐데 어떤 의미일지 이번 기회를 통해 느껴봤으면 좋겠습니다.

Commit:
refactor : 디텍토리 명 소문자로 컨벤션 일치

Commit:
refactor : assertThat 에 hasSize() 로 검증 변경

Commit:
refactor
1. carFactory 의 facotry 용도에 맞게 상태변수 제거
2. 제거된 상태에 맞추어서 createCars 메서드도 static 메서드로 변경

Commit:
refactor : static 메서드 호출방식으로 변경

Commit:
refactor :
1. Car 에 테스트용 위치를 조정하는 부생성자 추가
2. Car 시작을 0으로 수정하여 테스트 코드 예측값도 변경
3. Car에 테스트 위치 지정 방식에 맞춘 테스트 추가 및 변경

Commit:
recator :
1. ProgressRecord 의 타입을 record로 변환하여 불변성 보장
2. findWinners() 메서드 분리
3. ProgressRecord 의 불필요한 필드제거하여 Car로 일원화

Commit:
refacotr : 변경된 ProgressRecord 방식에 맞춰서 출력 및 일부로직 변경

Commit:
fix : ProgressRecord의 생성자에서 깊은 복사를 하는 방식으로 변경하여 똑같은 객체가 참조되어 마지막 이동기록이 똑같이 복사되는 문제 해결

Commit:
refactor:
1. 불필요한 GameResult 에 대한 List 감싸기를 제거
2. GameResult 내 불필요한 깊은복사 제거
3. GameFinalResult 를 GameResponse로 변경하여 UI용 객체임을 명시적으로 표시

Commit:
refactor
1. GameResult 가 모든 게임 결과를 상태로 가지고있으므로, 게임 결과 중 마지막을 기준으로 최종 승자 가리는것으로 역할지정

Commit:
refactor
1. GameMove 타입을 record로 하여 불변성을 보장
2. Car 의 name 필드의 불변성 보장

Commit:
refactor
1. 원시값을 포장하여 NonNegativeMoves 와 NonNegativeReadyCars 추가

Commit:
refactor
1. JoinCars 라는 car 들에 대한 일급컬렉션 생성하여 joinCars 중 1등을 추첨하는 기능 위치 이관
2. GameResult 에는 최종게임 상태에서 joinCars 호출

Commit:
refactor : NonNegativeMoves 와 NonNegativeReadyCars를 recored로 하여 불변성 보장

Commit:
doc :
목적 업데이트

Commit:
refactor
1. Car의 findLocation 사용을 최소화 하기 위해
비교 메서드 추가 및 Car을 인자로 받는 생성자 추가

javajigi commented on src/main/java/racingGame/business/CarFactory.java:
```diff
@@ -0,0 +1,19 @@
+package racingGame.business;
```
```suggestion
package racinggame.business;
```
패키지 이름은 전체가 소문자로 구현하는 것이 컨벤션

javajigi commented on src/main/java/racingGame/business/GameMove.java:
```diff
@@ -0,0 +1,26 @@
+    }
+    
+    private void nextMove(JoinCars joinCars) {
+        for(Car car: joinCars.cars()) {
```
joinCars에서 cars()로 데이터를 꺼내고 있다.
아래와 같이 joinCars에 메시지를 보내도록 구현해 보는 것은 어떨까?

`GameResult result = joinCars.race();`

객체를 너무 수동적인 존재로 바라보고 있는데 좀 더 능동적인 객체로 바라보는 의식적인 연습을 해볼 것을 추천

javajigi commented on src/main/java/racingGame/model/Car.java:
```diff
@@ -1,24 +1,68 @@
-    public Car() {
+    public Car(Car car) {
+        this(car.name, car.location);
+    }
```
굳이 필요한 생성자인가?

javajigi commented on src/main/java/racingGame/model/Car.java:
```diff
@@ -1,24 +1,68 @@
+        validate(name);
+        this.name = name;
         this.location = INIT_LOCATION;
+    }    
```
```suggestion
    public Car(String name) {
        this(name, INIT_LOCATION);
    }    
    
    public Car(String name, int location) {
       validate(name);
        this.name = name;
        this.location = location;
    }
```
모든 유효성 체크, 필드 값 초기화는 주 생성자에서..
부 생성자는 주 생성자를 호출

javajigi commented on src/main/java/racingGame/model/JoinCars.java:
```diff
@@ -0,0 +1,57 @@
+        return max;
+    }
+    
+    private JoinCars findMaxCar(int max) {
```
값을 List<Car>를 가지는 것은 같지만 내부 역할이 달라지는 것 아닐까?
JoinCars를 사용하기보다 Winners와 같은 다른 객체를 추가해 보는 것은 어떨까?

javajigi commented on src/test/java/racinggame/business/GameMoveTest.java:
```diff
@@ -0,0 +1,43 @@
+        int moveCount = 5;
+        String[] strings = {"pobi", "crong","honux"};
+        GameMove gameMove = new GameMove((new NonNegativeMoves(moveCount)));
+        JoinCars cars = createCars(new NonNegativeReadyCars(strings));
```
GameMove를 실행하기 위해 test fixture를 생성하는 단계가 너무 복잡함
test fixture가 복잡한 경우 리팩터링할 부분이 있을 것이라는 힌트로 생각해 보면 어떨까?

```java
GameMove gameMove = new GameMove({"pobi", "crong","honux"}, 5);

GameResult gameResult = gameMove.race();
```
test fixture가 복잡하면 GameMove를 사용해야 하는 다른 개발자들 또한 GameMove를 사용하는데 어려움이 있을 것으로 예상됨
위 코드와 같이 간단 명료하게 생성한 후 사용 가능하도록 구현해 보면 어떨까?

javajigi commented on src/test/java/racinggame/model/CarTest.java:
```diff
@@ -13,36 +14,68 @@ class CarTest {
+    @Test
+    void compareLocationTest() {
+        String name = "mo";
+        car = new Car(name,6);
```
```suggestion
        car = new Car(name, 6);
```
컨벤션 위반

javajigi commented on src/test/java/racinggame/model/GameResultTest.java:
```diff
@@ -0,0 +1,59 @@
+        JoinCars jc = new JoinCars(joinCars);
+        ProgressRecord progressRecord = new ProgressRecord(jc);
+        List<ProgressRecord> progressRecords = List.of(progressRecord);
+        GameResult gameResult = new GameResult(progressRecords);
```
우승자 구하는 로직을 ProgressRecord가 담당하면 어떨까?
race 마지막 ProgressRecord을 찾아 위임하면 되지 않을까?
GameResult에서 우승자 로직을 구현하다보니 위와 같이 복잡한 구조로 접근할 수 밖에 없음.

javajigi commented on src/test/java/racinggame/model/ProgressRecordTest.java:
```diff
@@ -4,39 +4,60 @@
+    }
+
+    @Test
+    void carNameTest() {
```
이 테스트는 굳이 필요할까?
단순히 값만 전달하고, 꺼내서 검증만 하는데 시간 투자대비 효과가 많이 떨어지지 않을까?

javajigi reviewed (changes requested):
4단계 미션 리팩터링하느라 수고했어요. 👍 
수업 중에 다룬 내용 포함해 미션 진행한 흔적이 여러 곳에서 보이네요.
특히 객체를 정말 작은 단위까지 분리하기 위해 노력한 흔적이 느껴지네요.
단, 여러 객체를 작은 단위로 분리했는데 객체 간의 협력을 어렵게 하다보니 복잡도가 많이 높아진 느낌이 들어요.
이렇게 과하게 분리하고 연결해 보는 것도 의미있다 생각해요.
다음 연습은 과하게 분리한 객체를 좀 더 단순화 해보는 연습을 해보면 좋겠습니다.

Commit:
refactor : 패키지 이름을 모두 소문자로 변경

Commit:
refactor
1. nextMove 를 joinCars 객체로 이관하고, 좀 더 능동적인 race로 이름 변경

Commit:
refactor
1. Car에 착각으로 반대로 배치한 주생성자, 부생성자 재구성
2. Car에 불필요한 생성자 제거

Commit:
refactor : 우승자 리스트를 가지는 역할을 하는 객체를 생성하여 일반적인 JoinCars와 역할분리

Commit:
refactor : 최종 1등을 뽑는 메서드를 전체 데이터를 가진 GameResult 객체가 아니라 ProgressRecord 객체로 이관하여 기록 및 1등 찾기에 간단하게 접근할 수있도록 변경

Commit:
refactor : testfixture 단계 단순화

Commit:
refactor : 과분리된 객체를 단순화위해 NonNegativeMoves 와 NonNegativeReadyCars 제거

Mo-bile commented:
안녕하세요 매번 리뷰 감사드립니다 👍 

몇가지 질문드려도 될까요?
# 질문
질문1. test fixture가 복잡하다고 하시는데, 이 복잡하다는 기준이 무엇인지 어렵습니다.
물론 특정 객체에게 능동성을 주지 못해서 복잡해지는 경우는 test fixture가 복잡하다는 것이 이해가 됩니다. 일부 수정한 경우도있구요!
음 그외에 복잡하다고 하시는 부분에서는 충분히 여기까지 필요한 데이터를 뽑아내기 위해 필요한게 아닌가? 이런생각이들어서요!
혹시 정량적인 기준이 있을까요?

질문2. 금번에 해주신 아래 피드백이 납득이 어렵습니다 좀 더 설명부탁드려도 될까요?
```
GameMove를 실행하기 위해 test fixture를 생성하는 단계가 너무 복잡함
test fixture가 복잡한 경우 리팩터링할 부분이 있을 것이라는 힌트로 생각해 보면 어떨까?

GameMove gameMove = new GameMove({"pobi", "crong","honux"}, 5);

GameResult gameResult = gameMove.race();
test fixture가 복잡하면 GameMove를 사용해야 하는 다른 개발자들 또한 GameMove를 사용하는데 어려움이 있을 것으로 예상됨
위 코드와 같이 간단 명료하게 생성한 후 사용 가능하도록 구현해 보면 어떨까?
```

이 부분의 경우 'GameMove' 에 `{"pobi", "crong","honux"}` 를 넣게되면 자동차 생성하는 역할을 GameMove에게 가게된다고 생각합니다.
이전 step 미션에서 move와 car가 다른 생명주기를 가져서 분리를 권하셨는데, 금번 리뷰에서는 `GameMove` 의 생성자에 다시 자동차 이름을 넣기를 권하셔서 혼란스럽습니다.

넣어야할 이유를 고민해도 두개의 역할은 다른데, 그리고 넣으면 GameMove안에서 `CarFactory.validateCar()` 클래스 메서드를 호출해서 생성을 맡겨야하나 이런생각했습니다. 그런데 다시 생각해도 이 부분은 더더욱 아닌 것 같습니다. 
혹시나 싶어 테스트를 위한 생성자를 만들라는 말씀이신가 싶어도, 테스트 용이라도 저기에 필드를 두는게 되나 이런생각도 들고요...

여러번 이렇게 저렇게 고민하다가 제가 혹시 말씀하신 의도를 잘못파악했나 싶기도 합니다.
혹시 좀 더 구체적으로 설명 부탁드려도 될까요?

질문3. 제 테스트 코드에 대한 의견에 대해서 어떻게 생각하실까요?
전반적인 피드백이 제 test fixture에 대한 내용이라고 생각합니다.
TDD가 단순히 구현이 아니라 적절하게 양을 가지고 통과하고 리팩터링한다고 생각합니다.
음 그런데 하면할수록 TDD가 정말로 쉽지 않음이 온몸으로 이해가 되네요
그리고 상위 객체일수록 테스트 픽스쳐가 당연히 많아진다고 생각합니다.
예컨데 processTest처럼요 이런 제 생각방식이 혹시 바람직한것과는 거리가 있을까요? ㅜㅜ

감사합니다

Commit:
refactor : 기존 Process와 GameMove를 인터페이스로 변경하여, 테스트에서는 자동차 움직임을 제어할 수있게끔 변경

javajigi commented on src/main/java/racinggame/business/impl/GameMove.java:
```diff
@@ -0,0 +1,30 @@
+    public List<ProgressRecord> moveCar(JoinCars joinCars) {
+        List<ProgressRecord> progressRecords = new ArrayList<>();
+        for(int i = 0; i < this.moves; i++) {
+            joinCars.race(progressRecords);
```
```suggestion
            ProgressRecord record = joinCars.race();
```
이와 같이 접근한느 것은 어떨까?

javajigi commented on src/main/java/racinggame/business/impl/GameProcess.java:
```diff
@@ -0,0 +1,17 @@
+import racinggame.model.*;
+import racinggame.response.GameResponse;
+
+public class GameProcess implements Process {
```
이번에 리팩터링하면서 여러 곳에 인터페이스를 추가함.
굳이 인터페이스가 필요할까? 오히려 오버엔지니어링이 된 것은 아닐까?
인터페이스는 어느 시점에 추가하는 것이 좋을까?

javajigi commented on src/main/java/racinggame/business/impl/TestMove.java:
```diff
@@ -0,0 +1,31 @@
+import racinggame.model.JoinCars;
+import racinggame.model.ProgressRecord;
+
+public record TestMove(int moves) implements Move {
```
GameMove와 중복 코드만 발생하고, 그리 좋은 접근 방식은 아닌 것으로 보임

javajigi commented on src/main/java/racinggame/model/JoinCars.java:
```diff
@@ -0,0 +1,65 @@
+        return findMaxCar(findMaxLocation());
+    }
+    
+    public void race(List<ProgressRecord> progressRecords) {
```
이 메서드와 같이 콜렉션을 전달해 값을 추가하기보다 race를 한 결과를 반환하는 형태로 구현해 볼 것을 추천

javajigi commented on src/main/java/racinggame/model/ProgressRecord.java:
```diff
@@ -0,0 +1,18 @@
+package racinggame.model;
+
+public record ProgressRecord(JoinCars joinCars) {
```
JoinCars을 전달한 후 다시 JoinCars을 생성하기보다 다음과 같이 구현하면 어떨까?

JoinCars에 `ProgressRecord toProcessRecord()` 메서드를 추가하고 ProgressRecord는 JoinCars가 아니라 List<Car>를 생성자의 인자로 받으면 어떨까?


```suggestion
public record ProgressRecord(List<Car> cars) {
```

앞의 List<ProgressRecord>을 전달하는 부분도 비슷함.
상태 값을 가지는 객체에 메시지를 보내 해당 객체가 현재 상태 값을 기반으로 ProgressRecord 같은 객체를 생성하도록 사고의 전환을 해보면 어떨까?
이렇게 구현했을 때 테스트 코드는 어떻게 달라지는지 고민해 볼 것을 추천

javajigi commented on src/main/java/racinggame/model/ProgressRecord.java:
```diff
@@ -0,0 +1,18 @@
+        this.joinCars = cars;
+    }
+    
+    public WinnerCars findGameWinners() {
```
생성자로 race가 끝난 자동차 목록(List<Car> cars)을 받는다면 우승자 구하는 로직이 ProgressRecord로 이동하지 않을까?

javajigi commented on src/main/java/racinggame/model/WinnerCars.java:
```diff
@@ -0,0 +1,22 @@
+    public WinnerCars(List<Car> cars) {
+        this.cars = new ArrayList<>(cars);
+    }
+    
```
자동차 이름을 꺼내와 List<String> carNames를 구하는 기능도 이 객체가 담당해야 하지 않을까?

javajigi commented on src/main/java/racinggame/GameApplication.java:
```diff
@@ -0,0 +1,13 @@
+    
+    public static void main(String[] args) {
+        Process process = new GameProcess();
+        ResultView.printAllResult(process.run(InputView.inputCarName(), InputView.inputMoveNum()));
```
현재 접근하는 방식이 너무 복잡하게 접근하는 것 같아 main() 메서드를 Controller 역할로 봤을 때의 대략적인 흐름을 볼 수 있는 코드 공유해 봄
지금까지 구현한 코드를 재사용하면서도 아래와 같은 구조로 변경하는 것이 어렵지 않을 것으로 생각함.
전체적인 흐름을 구현하는데 어려움을 겪고 있는 것 같아 MVC 기반의 샘플 예제를 제공해봄

```
public class RacingGameApplication {
    public static void main(String[] args) {
        String carNames = InputView.getCarNames();
        int tryNo = InputView.getTryNo();

        RacingGame racingGame = new RacingGame(carNames, tryNo);
        RacingResult result = null;
        while(racingGame.isEnd()) {
            result = racingGame.race();
            ResultView.printResult(result);
        }
        ResultView.printWinners(result); // RacingResult에서 우승자를 구해 출력
    }
}
```

위 코드에서 힌트를 얻어 본인의 코드에 개선할 부분이 있는지 검토해 보고 리팩토링할 부분이 있다면 리팩토링에 도전해 보세요.
단, 이미 위와 같은 구조로 구현했다면 특별히 개선할 부분은 없으면 바로 리뷰 요청해도 될 것 같아요.

Mo-bile replied:
@javajigi 
main() 이 컨트롤러 역할이라면
컨트롤러 레이어에서 
```java
while(racingGame.isEnd()) {
    result = racingGame.race();
    ResultView.printResult(result);
}
```
이 부분은 비즈니스 로직이라 생각이 되는데요.

저도 사실이렇게 생각했지만... 컨트롤러 레이어에서는 표현게층이기 때문에 오로지 결과에 대한 출력처리만 가능하다고 생각했었습니다
이또한 정답이 있는 부분은 아니겠지요?

저에게 보여주신 의도는 이렇게 쉽게 설계도 가능하다는것을 보여주시려는 의도 같습니다.. 맞으실까요?

javajigi replied:
@Mo-bile 
이 예시는 MVC 구조로 개발한다고 생각했을 때의 예시라 생각하면 됩니다.

"컨트롤러 레이어에서는 표현게층이기 때문에 오로지 결과에 대한 출력처리만 가능하다고 생각했었습니다."
저는 MVC에서 저는 "오로지 결과에 대한 출력처리만 가능하다"는 View가 담당해야 하는 역할이라 생각해요.
Controller는 Model과 View의 가교 역할로서 Model에 메시지를 보내 로직을 실행하도록 하고, 실행 결과를 반환받아 View에 전달하는 역할이라 생각합니다.
Controller가 이 같은 역할이라 생각하기 때문에 샘플로 보여준 코드와 같이 구현하는 것이 Controller의 역할이라 생각합니다.

javajigi commented on src/test/java/racinggame/business/MoveTest.java:
```diff
@@ -0,0 +1,67 @@
+        List<ProgressRecord> progressRecords = move.moveCar(cars);
+        
+        assertThat(progressRecords.getLast().joinCars().cars().getLast().findLocation())
+            .isEqualTo(moveCount);
```
테스트 값을 검증할 때 위와 같이 무한 get()을 통해 값을 꺼내와 값을 검증하는 것은 객체 설계에 복잡도가 높거나 의존 관계가 높다는 것을 반증함.
객체의 의존 관계를 끊어야할 힌트로 생각해 보면 좋겠음.

javajigi commented on src/test/java/racinggame/business/ProcessTest.java:
```diff
@@ -0,0 +1,41 @@
+        GameResponse gameResponse =  gp.run(carsNames, moves);
+        
+        String carsName = carsNames[carsNames.length - 1];
+        assertThat(gameResponse.winners().cars().getLast().getName()).isEqualTo(carsName);
```
테스트 값을 검증할 때 위와 같이 무한 get()을 통해 값을 꺼내와 값을 검증하는 것은 객체 설계에 복잡도가 높거나 의존 관계가 높다는 것을 반증함.
객체의 의존 관계를 끊어야할 힌트로 생각해 보면 좋겠음.

javajigi commented on src/test/java/racinggame/model/JoinCarsTest.java:
```diff
@@ -0,0 +1,49 @@
+    @Test
+    void findJointWinnersTest() {
+        List<Car> joinCars = new ArrayList<>();
+        joinCars.add(new Car("pobi",5));
```
```suggestion
        joinCars.add(new Car("pobi", 5));
```
컨벤션 위반

javajigi commented on src/test/java/racinggame/model/JoinCarsTest.java:
```diff
@@ -0,0 +1,49 @@
+        JoinCars jc = new JoinCars(joinCars);
+        
+        WinnerCars expectedWc = new WinnerCars(List.of(new Car("pobi",5), new Car("crong",5)));
+        assertThat(jc.findWinners()).isEqualTo(expectedWc);
```
👍

javajigi commented on src/test/java/racinggame/model/ProgressRecordTest.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.List;
+import org.junit.jupiter.api.Test;
+
+class ProgressRecordTest {
```
JoinCarsTest 에서 우승자에 대한 단위 테스트를 하고 있음.
ProgressRecordTest에서 우승자에 대한 단위 테스트를 진행하는 것은 중복 테스트로 보여짐
각 객체마다 책임과 역할이 있고, 해당 책임과 역할에 대해서만 테스트하면 됨.
이와 같이 JoinCarsTest의 테스트를 다시 ProgressRecordTest에서 진행하다보니 test fixture에 대한 복잡도도 높아짐
ProgressRecord를 생성하려면 여러 객체의 협력하에 생성해야 하기 때문에 test fixture가 복잡할 수 밖에 없음.

javajigi reviewed (changes requested):
피드백을 반영하기 위해 정말 여러가지 시도를 한 점이 인상적이네요. 👍 
질문 내용으로 보니 정말 고민이 많았을 것 같아요.

Q1. test fixture가 복잡하다고 하시는데, 이 복잡하다는 기준이 무엇인지 어렵습니다.
물론 특정 객체에게 능동성을 주지 못해서 복잡해지는 경우는 test fixture가 복잡하다는 것이 이해가 됩니다. 일부 수정한 경우도있구요!
음 그외에 복잡하다고 하시는 부분에서는 충분히 여기까지 필요한 데이터를 뽑아내기 위해 필요한게 아닌가? 이런생각이들어서요! 혹시 정량적인 기준이 있을까요?
A1. 정량적인 기준은 없습니다. test fixture가 복잡하다는 것은 테스트 대상이 되는 프로덕션 객체의 의존 관계가 복잡하거나 너무 많은 책임과 역할을 맡고 있기 때문일 가능성이 높습니다.
이 때는 테스트 코드를 개선보다 객체 설계를 개선하는 것이 테스트 코드의 복잡도를 낮출 수 있습니다.

Q2. 금번에 해주신 아래 피드백이 납득이 어렵습니다 좀 더 설명부탁드려도 될까요?
```
        GameMove gameMove = new GameMove((new NonNegativeMoves(moveCount)));
        JoinCars cars = createCars(new NonNegativeReadyCars(strings));
```
이 코드로 설명해 볼께요. 이와 같이 api를 사용할 때 매번 NonNegativeMoves, NonNegativeReadyCars을 생성해야만 api를 사용할 수 밖에 없다면 이 api를 쓰는 개발자 입장에서는 매번 객체로 포장해야하기 때문에 불편할 것 같아요.

객체로 포장한 api도 제공하겠지만 아래와 같이 쉽게 생성할 수 있도록 지원하면 좋지 않을까라 생각해요. 
```java
GameMove gameMove = new GameMove({"pobi", "crong","honux"}, 5);
GameResult gameResult = gameMove.race();
```

위와 같이 구현하면 GameMove() 생성자에는 NonNegativeMoves, NonNegativeReadyCars을 사용한 코드가 있겠죠. 코드를 구현할 때 객체 내부 구현에 많은 시간을 투자하고 복잡하면 이 객체를 사용하는 개발자는 편리성을 얻을 수 있지 않을까 생각하고 접근해 보면 어떨까요?

생성자 내부는 아래와 같은 구현이 있을 것 같네요.

```java
public GameMove(String names, int moveCount) {
   this(createCars(new NonNegativeReadyCars(strings)), new NonNegativeMoves(moveCount));
}

public GameMove(NonNegativeReadyCars cars, NonNegativeMoves moveCount) {
}
```

Q3. 제 테스트 코드에 대한 의견에 대해서 어떻게 생각하실까요?
A3. 답변이 될지 모르겠는데요. ProgressRecordTest에 피드백을 남기면서 제 생각을 전해봤어요. 피드백 확인해 보시고 궁금한 점 있으면 추가 질문 주세요.

Mo-bile committed:
refactor : List<ProgressRecord> 를 인자로 넣어서 통제하기 보다, 리턴하게해서 GameMove, JoinCars 두 객체 간 List<ProgressRecord> 에 대한 결합도를 줄임

Mo-bile committed:
refactor :
ProgressRecord 에 필드를 JoinCars에서 List<Car>로 변경해서 joinCars와의 강결합 방지,
그 이유는
1. ProgressRecord 에 JoinCars를 인자로 불러와서 다시 내부에서 deepCopy하는 것인데 joinCars 를 꺼내쓰기(파고들기)보다 joinCars에서 제공해주는 방향이 보다 내부 구조를 보여주지 않아서 결합도가 낮아짐
2. JoinCars 가 필드가 아니게 됨으로서 JoinCars 자체를 파고들어서 특정 행동을 취하지 않아도 되게 됨

Mo-bile committed:
refactor : WinnerCars 에 WinnerCarsName 을 구현하여 승자의 이름을 제공하는 행동 추가

Mo-bile committed:
refacotr : 컨벤션 준수

Mo-bile committed:
refactor : main메서드를 I/O 흐름이 보이게끔 리팩터링

Mo-bile committed:
refactor
1. 불필요한 Process 인터페이스 제거하여 오버엔지니어링 방지
2. Move인터페이스의 구현체 선택을 외부 주입방식으로 변경하여 process와 move간에 결합도 감소 및 process 테스트 시 GameMove없이도 테스트 가능

Mo-bile committed:
refactor
1. moveCar 의 응답타입을 변경하여 일급 컬렉션으로 변경하는 역할을 준다.

Mo-bile committed:
refactor : 호출 체이닝관련 리팩터링

Mo-bile commented:
안녕하세요?
우선 꼼꼼히 리뷰해주시고, 설명해주신점 감사드립니다 👍 
읽어보면서 많은점을 깨닫게되네요

이번에 알려주신 것을 이해하려다가 깨달은 점은
객체간의 협력(message)는 API간 통신이라는 점입니다.
뭔가 다른 개발자에게 API 문서를 전달한다고 생각하고 만드니 느낌은 오는것 같습니다.

그러면서도 하나의 객체에 다른객체에 대한 정보를 최소화하거나, get()등으로 파고들게 하지않아서 결합도를 낮추는것...
그리고 클래스의 상태를 최대한 이용해서 요청을 전달하기 등 다양한 방식으로 이해하려고 고민하고 고민한것 같습니다.

그러면서도 주신 피드백을 한번씩 살펴보면 솔직히 코에걸면 코걸이 귀에걸면 귀걸이 느낌인데,,, 이런느낌이 들수도있는것 맞을까요?
정답이 없는 부분이여서 그런것일까요? OOP, TDD는 참 녹록치 않은 것 같습니다 😂

몇가지 질문드려도 될까요?
1. Java Collection 을 일급 컬렉션으로 래핑 안한상태의 자료형은 어쩔수없이 get() 이 있어도 괜찮을까요?
무한 get() 으로 꺼내는 것에 대해서 리팩터링 포인트를 말씀주셔서 해보려고했는데요! 만약에 List<> 형태의 자료형이고 이것이 일급 컬렉션으로 되어 있지않다면 어쩔수없이 이용하는 getLast(), getFirst() 를 이용하는것은 충분히 감수가능할까요? 이러한 부분은 제외하고는 가급적 `ProgressRecord findLastRaceRecord()` 이런것도 만들어보면서 말씀주신 부분을 해결해보려고 시도했습니다.

2. `GameApplication.java` 관련 리뷰 부분에 MVC 예제 보여주신 것은 단순화한다면 어떻게할지 예시로서 보여주신 의도맞으실까요? 자세한건 해당부분에 댓글로 질문드렸습니다!

javajigi commented on src/main/java/racinggame/business/Process.java:
```diff
@@ -1,7 +1,21 @@
+        this.move = move;
+    }
+
+    public GameResponse run() {
+        JoinCars joinCars = CarFactory.createCars(names);
+        GameResult gameResult = move.moveCar(joinCars);
+        return new GameResponse(gameResult, gameResult.findLastRaceRecord().findGameWinners().WinnerCarsName());
+    }
```
```suggestion
    public GameResult run() {
        JoinCars joinCars = CarFactory.createCars(names);
        GameResult gameResult = move.moveCar(joinCars);
        return gameResult;
    }
```
위와 같이 GameResult만 반환해도 되지 않을까?
사용하는 곳에서는

```java
GameResult result = ...
ProcessRecord record = result.getProcessRecord(0//index);
ProcessRecord lastRecord = result.getLastRecord();
List<Car> winnersName =  lastRecord.winnersName();
```
이와 같이 접근해 가면 어떨까?

javajigi commented on src/main/java/racinggame/business/Process.java:
```diff
@@ -1,7 +1,21 @@
+    public GameResponse run() {
+        JoinCars joinCars = CarFactory.createCars(names);
+        GameResult gameResult = move.moveCar(joinCars);
+        return new GameResponse(gameResult, gameResult.findLastRaceRecord().findGameWinners().WinnerCarsName());
```
```suggestion
        return new GameResponse(gameResult, gameResult.findLastRaceRecord().findGameWinners().winnerCarsName());
```
메서드 이름은 소문자로 구현하는 것이 관례

javajigi reviewed (approved):
피드백 반영 하느라 수고했어요. 👍 
MVC 샘플 예제와 관련한 질문은 해당 질문에 남겼으니 확인해 보세요.

4단계는 지금 수준에서 merge할께요.
5단계 다음과 같이 연습해 보면 어떨까요?

- 현재까지 구현한 모든 코드를 삭제한다.
- 4단계 구현은 자동차 경주 게임이 끝날 때까지 모든 경주 기록을 기록하다가 마지막에 출력하는 방식으로 접근하고 있는데요. 
- 제가 MVC 예제에서 공유한 것처럼 매 시도횟수마다 경주 상태를 출력하는 방식으로 난이도를 좀 낮춰 구현해 보면 어떨까요? 이렇게 구현하면 객체 설계의 난이도가 좀 낮아질 수 있어서요. 새로운 인사이트를 얻을 수 있을 것 같아 제안해 봅니다.

Q1. Java Collection 을 일급 컬렉션으로 래핑 안한상태의 자료형은 어쩔수없이 get() 이 있어도 괜찮을까요?
무한 get() 으로 꺼내는 것에 대해서 리팩터링 포인트를 말씀주셔서 해보려고했는데요! 만약에 List<> 형태의 자료형이고 이것이 일급 컬렉션으로 되어 있지않다면 어쩔수없이 이용하는 getLast(), getFirst() 를 이용하는것은 충분히 감수가능할까요? 이러한 부분은 제외하고는 가급적 ProgressRecord findLastRaceRecord() 이런것도 만들어보면서 말씀주신 부분을 해결해보려고 시도했습니다.

A1. ProgressRecord findLastRaceRecord() 추가 시도 넘 좋네요.
일급 컬렉션으로 되어 있지 않다면 컬렉션 api를 쓰는 것은 어쩔 수 없습니다.
단, 재영님의 경우 여러 객체를 넘나들면서 계속 값을 꺼내는 방식으로 접근할 경우 객체 간의 의존성이 많아져(tightly coupling) 변경이 발생했을 때 취약할 수 있습니다. 
객체의 의존 관계를 느슨하게 구현(loosely coupling)하기 위한 의식적인 연습을 해볼 것을 추천해요.
